### PR TITLE
Fix bug where custom environment plugins were lost on dataset merge

### PR DESCRIPTION
**Pull request by williamcorsel**
### Summary

I ran into an issue where plugins loaded into the `Environment` are not retained after merging. Turns out a generator is being passed to [Environment.merge](https://github.com/openvinotoolkit/datumaro/blob/30b1add52d6fe458dba5e32e1f63ffeea999ad7b/src/datumaro/components/environment.py#L271) instead of the expected Sequence. This meant that custom plugins were never registered into the merged environment.

### How to test

Reproducible example:

```python
from datumaro.components.dataset import Dataset
from datumaro.components.environment import Environment
from datumaro.components.exporter import Exporter
from datumaro.components.hl_ops import HLOps
from datumaro.components.media import Image


class MyPlugin(Exporter):
    pass


environment = Environment()
environment.exporters.register('my_plugin', MyPlugin)

dataset1 = Dataset(media_type=Image, env=environment)
datasets = [dataset1, dataset1.clone()]
merged = HLOps.merge(*datasets)

assert 'my_plugin' in dataset1.env.exporters  # Passes
assert 'my_plugin' in merged.env.exporters  # Fails

```

### Checklist

- [x] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have added the description of my changes into [CHANGELOG](https://github.com/openvinotoolkit/datumaro/blob/develop/CHANGELOG.md).​
- [ ] I have updated the [documentation](https://github.com/openvinotoolkit/datumaro/tree/develop/docs) accordingly

### License

- [x] I submit _my code changes_ under the same [MIT License](https://github.com/openvinotoolkit/datumaro/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2024 Intel Corporation
#
# SPDX-License-Identifier: MIT
```

### CODE REVIEW

File: src/datumaro/components/hl_ops/__init__.py [hunk 1]
```diff
@@ -282,13 +282,14 @@ def merge(
...
-            # e.g., VocClassificationBase. this should be fixed and every object from
-            # Dataset.import_from should have "Dataset" type.
+            [
+                dataset.env
+                for dataset in datasets
+                if hasattr(dataset, "env")
+                # TODO: Sometimes, there is dataset which is not exactly "Dataset",
+                # e.g., VocClassificationBase. this should be fixed and every object from
+                # Dataset.import_from should have "Dataset" type.
+            ]
```
Comment (sooahleex) [inline src/datumaro/components/hl_ops/__init__.py:292]: I think those does not have to convert to `list` type. Do you have any reason to set those as list?

**Review by sooahleex — COMMENTED** [review]
Thank you for your interest to our project! Could you add an unit test for this?

File: src/datumaro/components/hl_ops/__init__.py (hunk 1 above)
Comment (williamcorsel) [reply src/datumaro/components/hl_ops/__init__.py:292]: Besides the fact that the type hint of the merge function defines a `Sequence`, the real issue is that [here](https://github.com/williamcorsel/datumaro/blob/57f176f149c86e8061ae94b789c2d4c4724298b1/src/datumaro/components/environment.py#L272) we loop over all items to check if they're default. If you supply a generator it means that the next loop [here](https://github.com/williamcorsel/datumaro/blob/57f176f149c86e8061ae94b789c2d4c4724298b1/src/datumaro/components/environment.py#L280) is always skipped.

**Review by sooahleex — APPROVED** [review]
LGTM. Thanks for your interest into our projects.

**Pull request merged**